### ISSUE TITLE
tf.nn.in_top_k strange behavior

### ISSUE THREAD

**Issue by vahidk** [tensorflow/tensorflow#13005]
Based on the documentation [here](https://www.tensorflow.org/api_docs/python/tf/nn/in_top_k) tf.nn.in_top_k treats ties as positive, which is almost never what you'd want. This is a major flaw, which leads to misleading metrics. An all zero predictions would give you 0 error with any targets. I think the default behavior should be to not count ties,

**Comment by reedwm** [comment]
@honkentuber does this proposal seem reasonable? We cannot change the default due to backwards compatibility concerns but we can add a flag.

**Comment by honkentuber** [comment]
+wicke,ispir

I don't know if a flag is worth the cost of complicating the interface, and
it doesn't address the original commenter's point that a counting ties is
unexpected or counterintuitive. It's probably worth updating the docs that
the more likelihood there is of ties, the less meaningful this metric is.

My intuition is that it should remain the way it is, or we should count
ties as a single true positive; i.e., in_top_k([10, 9, 8, 8, 8, 7, 6, 5],
k=3) should return either 5 or 3, but not 2.


Philip Tucker |  Software Engineer |  ptucker@google.com |

**Comment by facaiy** [comment]
If I understand correctly, for a sample, the sum of predict probabilities should be 1, right? So I cannot understand why all predicts can be zero.

**Comment by ppwwyyxx** [comment]
No one said they have to be probabilities. And even if they are probabilities, they can still be all equal.

This is actually a duplicate issue of #10767.

**Comment by honkentuber** [comment]
The likelihood of ties is much lower with continuous values. But you're
correct, this op does not assume that about the input.

**Comment by vahidk** [comment]
In my opinion there's no need for a flag. Counting ties as positive is never a wanted behavior for a metric.

**Comment by ppwwyyxx** [comment]
No one said this function has to be used for a metric..

**Comment by vahidk** [comment]
Are you sure? I think someone said.
https://www.tensorflow.org/versions/master/api_guides/python/nn#Evaluation

**Comment by ppwwyyxx** [comment]
I have no doubt they are designed for metrics but you can certainly use it for something else...

**Comment by martinwicke** [comment]
This is a duplicate of #10767. I wrote there that the flag would be good to clarify the behavior. This is more important for `top_k` than for `in_top_k`, but it may still be nice for consistency.
